src/libguac/encode-webp.c

```diff
@@ -0,0 +1,243 @@
+ *     picture->custom_ptr which contains the Guacamole stream writer structure.
+ *
+ * @return
+ *     True if writing was successful, false on failure.
+ */
+static int guac_webp_stream_write(const uint8_t* data, size_t data_size,
+        const WebPPicture* picture) {
+
+    guac_webp_stream_writer* const writer = (guac_webp_stream_writer*)picture->custom_ptr;
+    if (writer == NULL) {
+        return 0;
+    }
+
+    const unsigned char* current = data;
+    int length = data_size;
+
+    /* Append all data given */
+    while (length > 0) {
+
+        /* Calculate space remaining */
+        int remaining = sizeof(writer->buffer) - writer->buffer_size;
+
+        /* If no space remains, flush buffer to make room */
+        if (remaining == 0) {
+            guac_webp_flush_data(writer);
+            remaining = sizeof(writer->buffer);
+        }
+
+        /* Calculate size of next block of data to append */
+        int block_size = remaining;
+        if (block_size > length)
+            block_size = length;
+
+        /* Append block */
+        memcpy(writer->buffer + writer->buffer_size,
+               current, block_size);
+
+        /* Next block */
+        current += block_size;
+        writer->buffer_size += block_size;
+        length -= block_size;
+
+    }
+
+    return 1;
+}
+
+int guac_webp_write(guac_socket* socket, guac_stream* stream,
+        cairo_surface_t* surface, int quality) {
+
+    int width = cairo_image_surface_get_width(surface);
+    int height = cairo_image_surface_get_height(surface);
+    int stride = cairo_image_surface_get_stride(surface);
+    cairo_format_t format = cairo_image_surface_get_format(surface);
+
+    if (format != CAIRO_FORMAT_RGB24 && format != CAIRO_FORMAT_ARGB32) {
+        guac_error = GUAC_STATUS_INTERNAL_ERROR;
+        guac_error_message = "Invalid Cairo image format. Unable to create WebP.";
+        return -1;
+    }
+
+    /* Flush pending operations to surface */
+    cairo_surface_flush(surface);
+
+    unsigned char* data = cairo_image_surface_get_data(surface);
+
+    /* Configure WebP compression bits */
+    WebPConfig config;
+    if (!WebPConfigPreset(&config, WEBP_HINT_DEFAULT, quality)) return -1;  // version error
+    /* Add additional tuning: */
+    config.lossless = 0;
+    config.quality = quality;
+    config.thread_level = 1; /* Multi threaded */
+    config.method = 2; /* Compression method (0=fast/larger, 6=slow/smaller) */
+
+    WebPValidateConfig(&config);  /* verify parameter ranges */
+
+    /* Set up WebP picture and writer */
+    guac_webp_stream_writer writer;
+    WebPPicture picture;
+    WebPPictureInit(&picture);
+    picture.use_argb = 1;
+    picture.width = width;
+    picture.height = height;
+    WebPPictureAlloc(&picture);
+    picture.writer = guac_webp_stream_write;
+    picture.custom_ptr = &writer;
+    guac_webp_stream_writer_init(&writer, socket, stream);
+
+    /* Copy image data into WebP picture */
```

Comment:
I believe argb is allocated and owned by WebPPicture. Not sure if we can get around that.